[AArch64] Restrict .variant_pcs directive to ELF targets

MacDue opened the pull request:
Directive was implemented in c87bd2d8eb37 to support lazy binding and is emitted for vector PCS functions. It's specific to ELF but is currently emitted for all binary formats and crashing on non-ELF targets.

Fixes #138260

Co-authored-by: Cullen Rhodes <cullen.rhodes@arm.com>

MacDue committed:
[AArch64] Null check TargetStreamer before emitting .variant_pcs

This was causing crashes on Mach-O targets as we don't construct a
TargetStreamer for that object format. Other uses of the TargetStreamer
(TS) appear to be limited to ELF and COFF platforms (where it is none
null, so don't need changing).

sdesmalen-arm commented on llvm/test/CodeGen/AArch64/variant-pcs.ll:
```diff
@@ -2,6 +2,9 @@
 ; RUN:   | llvm-readobj --symbols - | FileCheck %s --check-prefix=CHECK-OBJ
 
+; Check we don't crash when using a Mach-O object format.
+; RUN: llc -mtriple=arm64-apple-macosx15.0.0 -mattr=+sve -filetype=obj -o /dev/null %s
```
I'm not entirely sure how useful this RUN line is, because it's guarding against a case that should never have happened.

MacDue replied:
Not sure I follow? All cases that cause a crash should never happen, but once fixed there should be a test to prevent a future regression.

sdesmalen-arm replied:
You're right, ignore my comment here, it seems like a valid case to test.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp:
```diff
@@ -1370,9 +1370,9 @@ void AArch64AsmPrinter::emitFunctionEntryLabel() {
-    auto *TS =
-        static_cast<AArch64TargetStreamer *>(OutStreamer->getTargetStreamer());
-    TS->emitDirectiveVariantPCS(CurrentFnSym);
+    if (auto *TS = static_cast<AArch64TargetStreamer *>(
```
Looking at this again, I think we should only emit the variant PCS attribute when the target's triple is ELF (`TM.getTargetTriple().isOSBinFormatELF()`) because this is specifically an ELF attribute. This means you can remove this condition, because for ELF targets it should never return a nullptr.

gbossu commented:
Curious: Why is it that the Mach-O format does not get a `TargetStreamer`? Does it really need no target-specific support?

c-rhodes committed:
Restrict .variant_pcs directive to ELF targets

Directive was implemented in c87bd2d to support lazy binding and is
emitted for vector PCS functions. It's specific to ELF but is currently
emitted for all binary formats and crashing on non-ELF targets.

Fixes #138260

[1]
https://github.com/ARM-software/abi-aa/blob/master/aaelf64/aaelf64.rst#st-other-values

MacDue committed:
Fixup test

MacDue commented:
> Curious: Why is it that the Mach-O format does not get a `TargetStreamer`? Does it really need no target-specific support?

Presumably not, there's currently only ELF and Windows COFF `TargetStreamer`s. The ELF one does not add much (seems like mainly a few attributes) and the Windows one seems to only handle various unwinding instructions. I guess Mach-O handles these differently.
